src/openapi/transport/queue.ts

```diff
@@ -0,0 +1,182 @@
+            authProvider.on(
+                authProvider.EVENT_TOKEN_RECEIVED,
+                this.authTokenReceived,
+                this,
+            );
+        }
+
+        this.transport = transport;
+    }
+
+    private tryEmptyQueue() {
+        if (
+            this.waitForPromises.length === 0 &&
+            (!this.authProvider || this.authProvider.getExpiry() > Date.now())
+        ) {
+            this.isQueueing = false;
+            this.emptyQueue();
+        }
+    }
+
+    prepareTransportMethod(method: HTTPMethodType) {
+        return (...args: HTTPMethodInputArgs) => {
+            if (!this.isQueueing) {
+                // checking expiry every time so that if device goes to sleep and is woken then
+                // we intercept a call about to be made and then do not have to cope with the 401 responses
+                if (
+                    this.authProvider &&
+                    this.authProvider.getExpiry() < Date.now()
+                ) {
+                    this.isQueueing = true;
+                    this.authProvider.refreshOpenApiToken();
+                }
+            }
+
+            const transportCallArguments = args;
+
+            return new Promise<OAPIRequestResult>((resolve, reject) => {
+                const queueItem = {
+                    method,
+                    args: transportCallArguments,
+                    servicePath: transportCallArguments[0] || '',
+                    urlTemplate: transportCallArguments[1] || '',
+                    urlArgs: transportCallArguments[2],
+                    options: transportCallArguments[3],
+                    resolve,
+                    reject,
+                };
+
+                if (this.isQueueing && this.shouldQueue(queueItem)) {
+                    this.addToQueue(queueItem);
+                } else {
+                    this.runQueueItem(queueItem);
+                }
+            });
+        };
+    }
+
+    private onWaitForPromiseResolved(promise: Promise<any>) {
+        this.waitForPromises.splice(this.waitForPromises.indexOf(promise), 1);
+
+        this.tryEmptyQueue();
+    }
+
+    private authTokenReceived = () => {
+        this.tryEmptyQueue();
+    };
+
+    waitFor(promise: Promise<any>) {
+        this.waitForPromises.push(promise);
+        this.isQueueing = true;
+        promise.then(() => this.onWaitForPromiseResolved(promise));
+    }
+
+    protected emptyQueue() {
+        for (let i = 0; i < this.queue.length; i++) {
+            this.runQueueItem(this.queue[i]);
+        }
+        this.queue.length = 0;
+    }
+
+    protected runQueueItem(item: QueueItem) {
+        this.transport[item.method](...item.args).then(
+            (...args: any[]) => {
+                item.resolve(...args);
+            },
+            // @ts-expect-error as it should return promise but returning direct void
+            (result: { status: HTTPStatusCode }, ...args: [any]) => {

```

Comment:
nitpick but should it be ReadonlyArray<any> ? currently it says there will only ever be one argument

Reply:
Looks like a mistake. Changed it to an array.